server/datastore/mysql/hosts.go

```diff
@@ -846,34 +846,39 @@ func (d *Datastore) SaveHostUsers(host *fleet.Host) error {
 		return err
 	}
 
-	incomingUsers := make(map[uint]bool)
+	keyForUser := func(u *fleet.HostUser) string { return fmt.Sprintf("%d\x00%s", u.Uid, u.Username) }
+	incomingUsers := make(map[string]bool)
+	var insertArgs []interface{}
 	for _, u := range host.Users {
-		incomingUsers[u.Uid] = true
-
-		if _, err := d.writer.Exec(
-			`INSERT IGNORE INTO host_users (host_id, uid, username, user_type, groupname) VALUES (?, ?, ?, ?, ?)`,
-			host.ID, u.Uid, u.Username, u.Type, u.GroupName,
-		); err != nil {
-			return errors.Wrap(err, "insert users")
-		}
+		insertArgs = append(insertArgs, host.ID, u.Uid, u.Username, u.Type, u.GroupName)
+		incomingUsers[keyForUser(&u)] = true
 	}
 
 	var removedArgs []interface{}
 	for _, u := range currentHost.Users {
-		if _, ok := incomingUsers[u.Uid]; !ok {
-			removedArgs = append(removedArgs, u.ID)
+		if _, ok := incomingUsers[keyForUser(&u)]; !ok {
+			removedArgs = append(removedArgs, u.Username)
 		}
 	}
 
+	insertValues := strings.TrimSuffix(strings.Repeat("(?, ?, ?, ?, ?),", len(host.Users)), ",")
+	insertSql := fmt.Sprintf(
+		`INSERT IGNORE INTO host_users (host_id, uid, username, user_type, groupname) VALUES %s`,
+		insertValues,
+	)
+	if _, err := d.writer.Exec(insertSql, insertArgs...); err != nil {
+		return errors.Wrap(err, "insert users")
+	}
+
 	if len(removedArgs) == 0 {
 		return nil
 	}
 	removedValues := strings.TrimSuffix(strings.Repeat("?,", len(removedArgs)), ",")
 	removedSql := fmt.Sprintf(
-		`UPDATE host_users SET removed_at = CURRENT_TIMESTAMP WHERE id IN (%s)`,
+		`UPDATE host_users SET removed_at = CURRENT_TIMESTAMP WHERE host_id = ? and username IN (%s)`,
```

Comment:
I see that we use soft deletes, but what happens if a deleted user is re-introduced for the same host? It looks like it would end up being the same primary key (say, if the `uid` is also the same), so INSERT IGNORE would ignore it as it already exists (but removed), and it would stay that way?

Reply:
Correct! I realized that yesterday, and that's why the second PR.